### PR TITLE
Tutorial on linear mixed fluid-structure interaction system.

### PR DESCRIPTION
**Pull request by mmtjs**
Demonstrates use of subdomain IDs from the mesh and use of DirichletBC to restrict solvers to appropriate regions of the whole domain.

### CODE REVIEW

File: demos/linear_mixed_system/linear_mixed_system.py.rst [hunk 1]
```diff
@@ -0,0 +1,328 @@
+Linear mixed fluid-structure system
```
Comment (dham) [inline demos/linear_mixed_system/linear_mixed_system.py.rst]: This is a perfectly good title, however the name you are using for directories and files here is confusing. "linear_mixed_system" could refer to a huge range of finite element problems. Please use a name which indicates that this is fluid structure interaction.

File: demos/linear_mixed_system/linear_mixed_system.py.rst [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+
+.. rst-class:: emphasis
+
+This tutorial is aimed to demonstrate the use of subdomain functionality and show how to describe a system consisting of multiple materials in Firedrake.
+
+The tutorial was contributed by `Tomasz Salwa <mailto:mmtjs@leeds.ac.uk>`__
+and `Onno Bokhove <mailto:O.Bokhove@leeds.ac.uk>`__.
+
+The work is based on the article "Variational modelling of wave-structure interactions with an
+offshore wind-turbine mast" by Tomasz Salwa, Onno Bokhove and Mark Kelmanson :cite:`Salwa:2017`. The authors gratefully acknowledge funding from European Commission, Marie Curie Actions - Initial Training Networks (ITN), project number 607596.
```
Comment (dham) [inline demos/linear_mixed_system/linear_mixed_system.py.rst]: Please move lines 12 and 13 to the bottom of the page. They are not the first thing that a reader needs to see.

File: demos/linear_mixed_system/linear_mixed_system.py.rst [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+
+The model considered consists of fluid with a free surface and an elastic solid. We will be using interchangeably notions of fluid/water and structure/solid/beam. For simplicity (and speed of computation) we consider a model in 2D, however it can be easily generalized to 3D. The starting point is the linearized version (domain is fixed) of the fully nonlinear variational principle. In non-dimensional units:
+
+.. math::
+
+    0 = & \delta \int_0^{t_{\text{end}}} \int \left( \partial_t{\eta} \phi - \frac{1}{2} \eta^2 \right) {\mathrm d} S_f - \int \frac{1}{2} |\nabla \phi|^2 {\mathrm d} V_F \\
+    & + \int {\bf n} \cdot \partial_t {\bf X} \phi \, {\mathrm d} S_s\\
+    & + \int \rho_0 \partial_t {\bf X} \cdot {\bf U} - \frac 12 \rho_0 |{\bf U}|^2 - \frac 12 \lambda e_{ii}e_{jj} - \mu e_{ij}^2\, {\mathrm d} V_S \, {\mathrm d} t \, ,
+
```
Comment (dham) [inline demos/linear_mixed_system/linear_mixed_system.py.rst]: Please use dx for volume integrals and ds, dS for interior and exterior surface integrals in order to be consistent with UFL and the rest of the site.

File: demos/linear_mixed_system/linear_mixed_system.py.rst [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+    \\
+    & \int \nabla v \cdot \nabla \phi^{n+1} \, {\mathrm d} V_F \underline{ - \int v {\bf n} \cdot {\bf U}^{n+1} \, {\mathrm d} S_s }= 0 \\ %\hspace{1cm} (+ \text{Dirichlet BC at } \partial \Omega_f)\\
+    & \int v \eta^{n+1} \, {\mathrm d} S_f = \int v \eta^n \, {\mathrm d} S_f + \Delta  t \int \nabla v \cdot \nabla \phi^{n+1} \, {\mathrm d} V_F   \underline{ - \Delta t \int v {\bf n} \cdot {\bf U}^{n+1}\, {\mathrm d} S_s }\\
+    & \int {\bf v} \cdot {\bf X}^{n+1} \, {\mathrm d} V_S = \int {\bf v} \cdot ( {\bf X}^n + \Delta t {\bf U}^{n+1} ) \, {\mathrm d} V_S \, .
+    \end{align}
+
+The underlined terms are the coupling terms. Note that the first equation for :math:`\phi` at the free surface is solved on the free surface only, the last equation for :math:`{\bf X}` in the structure domain, while others in both domains. Moreover, the second and third equations for :math:`\phi` and :math:`{\bf U}` need to be solved simultaneously. Geometry of the system with initial condition is shown below.
+
```
Comment (dham) [inline demos/linear_mixed_system/linear_mixed_system.py.rst]: "while others in both domains" is ungrammatical. Perhaps "while the others are solved in bot domains".

Comment (dham) [inline demos/linear_mixed_system/linear_mixed_system.py.rst]: "Geometry of the system" needs to be "The geometry of the system"

File: demos/linear_mixed_system/linear_mixed_system.py.rst [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+    dt = 0.005 # time step [s]
+    g = 9.8 # gravitational acceleration
+    # water
+    Lx = 20. # length of the tank [m] in x-direction; needed for computing initial condition
+    Lz = 10. # height of the tank [m]; needed for computing initial condition
+    rho = 1000. # fluid density in kg/m^2 in 2D [water]
+    # solid parameters
+    rho_B = 7700. # structure density in kg/m^2 in 2D [steel]
+    lam = 1e7 # 9.695e10 for steel, # N/m in 2D - first Lame constant
+    mu = 1e7 # 7.617e10 for steel, # N/m in 2D - second Lame constant
```
Comment (dham) [inline demos/linear_mixed_system/linear_mixed_system.py.rst]: This reads slightly strangely. You are using values of 1e7 (which I assume is correct) but then you give an explanation for a different value.

File: demos/linear_mixed_system/linear_mixed_system.py.rst [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+This tutorial is aimed to demonstrate the use of subdomain functionality and show how to describe a system consisting of multiple materials in Firedrake.
+
+The tutorial was contributed by `Tomasz Salwa <mailto:mmtjs@leeds.ac.uk>`__
+and `Onno Bokhove <mailto:O.Bokhove@leeds.ac.uk>`__.
+
+The work is based on the article "Variational modelling of wave-structure interactions with an
+offshore wind-turbine mast" by Tomasz Salwa, Onno Bokhove and Mark Kelmanson :cite:`Salwa:2017`. The authors gratefully acknowledge funding from European Commission, Marie Curie Actions - Initial Training Networks (ITN), project number 607596.
+
+The model considered consists of fluid with a free surface and an elastic solid. We will be using interchangeably notions of fluid/water and structure/solid/beam. For simplicity (and speed of computation) we consider a model in 2D, however it can be easily generalized to 3D. The starting point is the linearized version (domain is fixed) of the fully nonlinear variational principle. In non-dimensional units:
+
```
Comment (dham) [inline demos/linear_mixed_system/linear_mixed_system.py.rst]: Please use the more common UK spelling convention of -ise throughout (e.g. generalised).

File: demos/linear_mixed_system/linear_mixed_system.py.rst [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+And last, mixed functions in the mixed domain::
+
+    trial_f, trial_s = TrialFunctions(mixed_V)
+    v_f, v_s = TestFunctions(mixed_V)
+    tmp_f = Function(V_W)
+    tmp_s = Function(V_B)
+    result_mixed = Function(mixed_V)
+
+We need auxiliary indicator functions, that are 0 in one subdomain and 1 in the other. They are needed both in "CG" and "DG" space. We use the fact, that the fluid and structure subdomains are defined in the mesh file with an appropriate ID number that Firedrake is able to recognize. That can be used in constructing indicator functions::
+
```
Comment (dham) [inline demos/linear_mixed_system/linear_mixed_system.py.rst]: Remove the comma after fact

File: demos/linear_mixed_system/linear_mixed_system.py.rst [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+        phi.assign(tmp_f)
+        U.assign(tmp_s)
+        LVS_eta.solve()
+        tmp_f, _ = result_mixed.split()
+        eta.assign(tmp_f)
+        LVS_X.solve()
+
+        output_data()
+
+The result of the computation, visualised with `paraview <http://www.paraview.org/>`__, is shown below.
```
Comment (dham) [inline demos/linear_mixed_system/linear_mixed_system.py.rst]: What would be really cool at this point would be a youtube video of the solution. You can see how to embed videos by looking in `docs/source/index.rst`.

**Comment by dham** [comment]
Please note that you also need to modify `documentation.rst` to add your demo to the list of advanced tutorials at the bottom of the page.

**Comment by mmtjs** [comment]
Updated according to comments.

**Pull request merged**